### ISSUE TITLE
Invalid format exception when reading results of code search, but only on linux

### ISSUE THREAD

**Issue by Blecki** [octokit/octokit.net#684]
I'm encountering an AggregateException which contains a single InvalidFormat exception when attempting to read the results of a code search. The exception is only thrown when running on linux. The code does as expected when built and run from visual studio, and when built with xamarin on a mac. I've been able to replicate the problem inside a C# repl running on the linux server. The server is running Ubuntu 14.04.1.

```
var githubClient = new Octokit.GitHubClient(new Octokit.ProductHeaderValue("Reddit-Mud"));
var codeSearch = new Octokit.SearchCodeRequest(".cs") {
                    Repo = "Reddit-Mud/RMUD-DB",
                    In = new[] { Octokit.CodeInQualifier.Path },
                    Page = 1
                };
var codeResult = githubClient.Search.SearchCode(codeSearch);
var result = codeResult.Result; // <- Throws exception.
```
The actual project is here https://github.com/Reddit-Mud/RMUD , the original code that failed is here https://github.com/Reddit-Mud/RMUD/blob/master/RMUD/GithubDatabase/EnumerateDatabase.cs

**Comment by shiftkey** [comment]
@Blecki do you know what version of Mono you're running?

**Comment by Trevoke** [comment]
```bash
$ mono --version
Mono JIT compiler version 3.2.8 (Debian 3.2.8+dfsg-4ubuntu1)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	LLVM:          supported, not enabled.
	GC:            sgen
```

Running on Ubuntu 14.04.1 LTS -- a very recent image from Digital Ocean.

**Comment by Trevoke** [comment]
By the way, if this wasn't clear, I'm working with Blecki on this, so what I pasted is what Blecki's version is.

**Comment by shiftkey** [comment]
This appears to be an issue how we're using HttpClient on Linux (still occurring in 0.6.2):

```
System.FormatException: Invalid format.

at System.Net.Http.Headers.HttpHeaders.AddInternal (System.String name, IEnumerable`1 values, System.Net.Http.Headers.HeaderInfo headerInfo, Boolean ignoreInvalid) [0x00000] in <filename unknown>:0
at System.Net.Http.Headers.HttpHeaders.Add (System.String name, IEnumerable`1 values) [0x00000] in <filename unknown>:0
at System.Net.Http.Headers.HttpHeaders.Add (System.String name, System.String value) [0x00000] in <filename unknown>:0
at Octokit.Internal.HttpClientAdapter.BuildRequestMessage (IRequest request) [0x00000] in <filename unknown>:0 
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <filename unknown>:0
at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[Octokit.IResponse`1[Octokit.SearchCodeResult]].GetResult () [0x00000] in <filename unknown>:0
at Octokit.Connection+<RunRequest>d__1f`1[Octokit.SearchCodeResult].MoveNext () [0x00000] in <filename unknown>:0
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <filename unknown>:0
at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[Octokit.IResponse`1[Octokit.SearchCodeResult]].GetResult () [0x00000] in <filename unknown>:0
at Octokit.Connection+<Run>d__1b`1[Octokit.SearchCodeResult].MoveNext () [0x00000] in <filename unknown>:0
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <filename unknown>:0
at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[Octokit.IResponse`1[Octokit.SearchCodeResult]].GetResult () [0x00000] in <filename unknown>:0
at Octokit.ApiConnection+<Get>d__0`1[Octokit.SearchCodeResult].MoveNext () [0x00000] in <filename unknown>:0
```

I now have a proper dev environment, let's see if I can isolate where exactly it's falling over.

**Comment by shiftkey** [comment]
So I've settled on the simplest repro to trigger this:

```csharp
var message = new HttpRequestMessage (
    HttpMethod.Get,
    "https://api.github.com/search/code?page=1&per_page=100&order=desc&q=.cs+in:path"
);

message.Headers.Add ("Accepts", "application/vnd.github.v3+json; charset=utf-8");
// and this is the line that blows up
message.Headers.Add ("User-Agent", "ha-ha-mono (Unix 3.13.0; amd64; en-US; Octokit 0.6.3)");
```

Works on Windows, so likely an implementation issue with `System.Net.Http.Headers.HttpRequestHeaders` on Mono.

:cool:

**Comment by shiftkey** [comment]
I'll just leave this here: https://github.com/mono/mono/blob/5728c1a3dd946119e566f7afe7bf79b85f82f89d/mcs/class/System.Net.Http/System.Net.Http.Headers/HttpHeaders.cs#L179-L223

**Comment by shiftkey** [comment]
As a workaround until we can get a fix into Mono proper, we could use `TryAddWithoutValidation` for the User-Agent header value:

```
message.Headers.TryAddWithoutValidation("User-Agent", "ha-ha-mono (Unix 3.13.0; amd64; en-US; Octokit 0.6.3)");
```

Thoughts @haacked?

**Comment by akoeplinger** [comment]
@shiftkey your repro works fine on the latest Mono 3.12 release.
@Trevoke can you try updating Mono as shown here: http://www.mono-project.com/docs/getting-started/install/linux/

**Comment by shiftkey** [comment]
@akoeplinger thanks for confirming it's been fixed before I went stepping through the source :heart:

**Comment by thedillonb** [comment]
I just upgraded to the 0.7.0 release and started seeing this issue. I am (apparently) running Mono 3.12.0 but still seeing this issue.

![screen shot 2015-02-23 at 8 28 32 pm](https://cloud.githubusercontent.com/assets/1644068/6341628/a169c480-bb9a-11e4-8677-0fcb13477834.png)

**Comment by shiftkey** [comment]
@thedillonb could you try this script and see if it's not a different problem?

```
var message = new HttpRequestMessage (
    HttpMethod.Get,
    "https://api.github.com/"
);

message.Headers.Add ("Accepts", "application/vnd.github.v3+json; charset=utf-8");
// and this is the line that blows up
message.Headers.Add ("User-Agent", "ha-ha-mono (Unix 3.13.0; amd64; en-US; Octokit 0.7.0)");
```

@akoeplinger thoughts?

**Comment by akoeplinger** [comment]
@thedillonb yep, I see the same on Mono 3.12 and Mono master with you repro code, so it looks like @shiftkey's repro code didn't fully get to the root of the problem :cry: Would be interesting to know which header value causes this so the Mono guys can fix it.

**Comment by shiftkey** [comment]
I can re-test this against Mono 3.12 and dig into it further. Gimme a day or so.

**Comment by thedillonb** [comment]
@shiftkey, I just ran your repro code without problem on my setup. Perhaps this is a different issue?

I just build the project from master, created a dummy console project and referenced the **Octokit-Portable.dll** - no Issue. If I reference the **Octokit-Mono.dll** then I have a bad time :-1:

**Comment by shiftkey** [comment]
@thedillonb yep, this is a different issue:

```
var request = new HttpRequestMessage ();
request.Headers.Add ("Accept", "application/vnd.github.moondragon+json; charset=utf-8,application/vnd.github.v3+json; charset=utf-8");
```

This was a change in #709 and the `,` in here appears to be the issue (space or no space after doesn't make a difference). cc @akoeplinger 

> If I reference the Octokit-Mono.dll then I have a bad time :-1:

That makes sense - Octokit-Portable references [a different System.Net.Http](https://github.com/octokit/octokit.net/blob/master/Octokit/Octokit-Portable.csproj#L376-L384) to the one in Octokit-Mono

**Comment by akoeplinger** [comment]
@shiftkey I filed a Xamarin bug for this: https://bugzilla.xamarin.com/show_bug.cgi?id=27352

Note that you can workaround this by just split Add() into multiple calls (this works on Mono 3.12 as well):
```
request.Headers.Add ("Accept", "application/vnd.github.moondragon+json; charset=utf-8");
request.Headers.Add ("Accept", "application/vnd.github.v3+json; charset=utf-8");
```

**Comment by thedillonb** [comment]
Looks like it has been fixed in Mono master.

@akoeplinger, in your example above where you are just using two `Add` calls, what is the type of `request.Headers`? Isn't it a `Dictionary<string, string>` thus arn't you just overriding the first with the second?

I'm getting away with just `request.Headers["Accept"] = "application/vnd.github.v3+json; charset=utf-8";` in a custom `IHttpClient` implementation for now.

**Comment by akoeplinger** [comment]
@thedillonb it isn't overriding, multiple Add() calls are fine. The type is [HttpRequestHeaders](https://github.com/mono/mono/blob/master/mcs/class/System.Net.Http/System.Net.Http.Headers/HttpRequestHeaders.cs).

**Comment by shiftkey** [comment]
I've fixed this issue incidentally by shipping #734 which dropped the multiple `Accept` headers. When I get around to addressing how we handle preview APIs properly, I'll keep this workaround in mind.
